File: cpp_sdk/cpp_dist_bundle/benchmarks/observation.cpp
#include "observation.h"

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <fstream>
#include <numeric>
#include <ostream>
#include <sstream>

namespace Trueface {
namespace Benchmarks {

void Observation::emitToUser() {
    // screen output for reporting progress to user
    std::cout << "Average time " << m_benchmarkName;
    if (!m_benchmarkSubType.empty()) {
        std::cout << " (" << m_benchmarkSubType << ")";
    }

    auto precision{std::cout.precision()};
    std::cout << ": " << std::fixed << std::setprecision(3) << m_time.mean << " ms"
        << std::defaultfloat << std::setprecision(precision);

    if (m_params.batchSize > 1) {
        std::cout << " | batch size = " << m_params.batchSize;
    }

    std::cout << " | " << m_params.numIterations << " iterations " << std::endl;
}

TimeResult summarizeTimes(const Parameters &params, std::vector<float> times) {
    //
    // times should be passed in NANOSECONDS as milliseconds do not have the resolution for
    // some individual executions. So, the following divide through by 1000 to convert to ms
    // for consumption down stream AFTER the calculations are done.
    //
    std::transform(times.begin(), times.end(), times.begin(), [&params](float val) {
        return val / static_cast<float>(params.batchSize);
    });
    auto total = std::accumulate(times.begin(), times.end(), 0.0f);
    auto mean = total / params.numIterations;
    const auto minmax = std::minmax_element(times.begin(), times.end());

    const size_t sz = times.size();
    auto variance_func = [&mean, &sz](float accumulator, const float &val) {
        return accumulator + ((val - mean) * (val - mean) / (sz - 1));
    };
    auto variance = std::accumulate(times.begin(), times.end(), 0.0f, variance_func) / 1000.f;

    constexpr float nsPerMs{1000.f * 1000.f};
    return TimeResult{
        total / nsPerMs,
        mean / nsPerMs,
        variance / (nsPerMs * 1000.f),
        *minmax.first / nsPerMs,
        *minmax.second / nsPerMs};
}

Observation::Observation(const std::string &version, bool isGpuEnabled,
                const std::string &benchmarkName, const std::string &benchmarkSubType,
                const Parameters &params, const std::vector<float> &times,
                float memoryUsage)
    : m_version{version}, m_isGpuEnabled{isGpuEnabled}, m_benchmarkName{benchmarkName},
    m_benchmarkSubType{benchmarkSubType}, m_params{params},
    m_time{summarizeTimes(params, times)}, m_memoryUsage{memoryUsage} {
    emitToUser();
}

std::ostream& operator<<(std::ostream& out, const Observation& o) {
    const auto& params = o.getParameters();
    const auto& time = o.getTimeResult();
    auto precision{out.precision()};
    out << o.getVersion() << ", "
        << (o.getIsGpuEnabled() ? "GPU" : "CPU") << ","
        << "\"" << o.getBenchmarkName() << "\","
        << "\"" << o.getBenchmarkSubType() << "\","
        << params.batchSize << ","
        << params.numIterations << ","
        << std::fixed << std::setprecision(3)
        << time.total << "," << time.mean << "," << time.variance << ","
        << time.low << "," << time.high << ","
        << o.getMemoryUsage()
        // reset iomanip
        << std::defaultfloat << std::setprecision(precision);

    return out;
}

ObservationCSVWriter::ObservationCSVWriter(const std::string& path)
    : m_path{path}, m_writeHeaders{!doesFileExist(path)} {

}

void ObservationCSVWriter::write(const ObservationList &observations) {
    // write observations to a csv
    std::ofstream out{m_path, std::ios::app};
    if (m_writeHeaders) {
        out << "SDK Version, "
            << "GPU or CPU, "
            << "Benchmark Name, "
            << "Benchmark Type or Model, "
            << "Batch Size, " << "Number of Iterations, "
            << "Total Time (ms), Mean Time (ms), Variance (ms), Low (ms), High (ms), "
            << "Memory Usage (MB)"
            << "\n";
    }

    for (const auto& observation : observations) {
        out << observation << "\n";
    }

    out.close();
}

bool ObservationCSVWriter::doesFileExist(const std::string& path) {
    std::ifstream f{m_path};
    return f.good();
}


} // namespace Benchmarks
} // namespace Trueface

File: python/python_dist_bundle/benchmarks/observation.py
from dataclasses import dataclass
import statistics
from typing import List
import utils


@dataclass
class TimeResult:
    total: float
    mean: float
    variance: float
    low: float
    high: float


@dataclass
class Observation:
    def __init__(self, version: str, is_gpu_enabled: bool, benchmark_name: str, benchmark_subtype: str, params: utils.Parameters, times: List[float]):
        self.version = version
        self.is_gpu_enabled = is_gpu_enabled
        self.benchmark_name = benchmark_name
        self.benchmark_subtype = benchmark_subtype
        self.params = params
        self.time = self._summarize_times(params, times)
        print(self)

    def _summarize_times(self, params: utils.Parameters, times: List[float]) -> TimeResult:
        normalized_times = list(map(lambda x: x / params.batch_size, times))

        time_result = TimeResult(
            total=sum(normalized_times) / utils.Stopwatch.ns_in_ms,
            mean=statistics.fmean(normalized_times) / utils.Stopwatch.ns_in_ms,
            low=min(normalized_times) / utils.Stopwatch.ns_in_ms,
            high=max(normalized_times) / utils.Stopwatch.ns_in_ms,
            variance=statistics.variance(normalized_times) / utils.Stopwatch.ns_in_ms)

        return time_result

    def __str__(self) -> str:
        s = f'Average time {self.benchmark_name}'
        if self.benchmark_subtype:
            s += f' ({self.benchmark_subtype})'
        s += f': {self.time.mean: 0.3f} ms'

        if self.params.batch_size > 1:
            s += f' | batch size = {self.params.batch_size}'

        s += f' | {self.params.num_iterations} iterations'

        return s
